### ISSUE TITLE
Module is trying to pull info from a docker container that doesn't exist

### ISSUE THREAD

**Issue by melissacrawford396** [monitoringartist/Zabbix-Docker-Monitoring#28]
I am using Zabbix to monitor docker containers for Mesos. There are three healthy containers that are reporting correctly, but there is a fourth one listed on the display even though only 3 containers exist.

I looked into the logs and these are the errors each time the information is pulled for the 4th non-existent-mystery container:
`4469:20151029:133848.468 Cannot open Docker container metric file: '/sys/fs/cgroup/cpu,cpuacct/system.slice/docker-system.slice.scope/cpuacct.stat'
  4469:20151029:133848.468 active check "docker.cpu[system.slice,system]" is not supported
  4469:20151029:133848.468 Cannot open Docker container metric file: '/sys/fs/cgroup/cpu,cpuacct/system.slice/docker-system.slice.scope/cpuacct.stat'
  4469:20151029:133848.468 active check "docker.cpu[system.slice,user]" is not supported
  4469:20151029:133848.468 Cannot open Docker container metric file: '/sys/fs/cgroup/memory/system.slice/docker-system.slice.scope/memory.stat'
  4469:20151029:133848.468 active check "docker.mem[system.slice,total_cache]" is not supported
  4469:20151029:133848.469 Cannot open Docker container metric file: '/sys/fs/cgroup/memory/system.slice/docker-system.slice.scope/memory.stat'
  4469:20151029:133848.469 active check "docker.mem[system.slice,total_rss]" is not supported
  4469:20151029:133848.469 Cannot open Docker container metric file: '/sys/fs/cgroup/memory/system.slice/docker-system.slice.scope/memory.stat'
  4469:20151029:133848.469 active check "docker.mem[system.slice,total_swap]" is not supported`

From looking through the code in zabbix_module_docker.c, it looks like the **container** variable seems to be set to **system.slice** instead of an actual docker container id. 

`'/sys/fs/cgroup/memory/system.slice/docker-**system.slice**.scope/memory.stat'`

In the **zbx_module_docker_get_fci** function that is getting the container value within **zbx_module_docker_up** function, it is only checking if it starts with a '/' not if the container id is a number. 

` char*  zbx_module_docker_get_fci(char *fci)
{
        zabbix_log(LOG_LEVEL_DEBUG, "In zbx_module_docker_get_fci()");

        // if first character is '/', then detect fci
        if (fci[0] != '/')
        {
            zabbix_log(LOG_LEVEL_DEBUG, "Original full container id will be used");
            return zbx_strdup(NULL, fci);
        } `

`container = zbx_module_docker_get_fci(get_rparam(request, 0));`

**So, the question:
Am I missing something? There were some docker containers that had failed earlier. Would that cause this kind of issue on our display?**

<img width="513" alt="screen shot 2015-10-29 at 4 26 33 pm" src="https://cloud.githubusercontent.com/assets/8278766/10831198/afeccd80-7e5a-11e5-89ae-37548ddfca2e.png">
*this last entry is what I'm talking about ^^*

**Comment by jangaraj** [comment]
Your containers haven't been discovered correctly (last entry shouldn't be there). Could you please provide outputs:

    ls -lah /sys/fs/cgroup/memory/system.slice/
    ls -lah /sys/fs/cgroup/cpu,cpuacct/system.slice/
    zabbix_agentd -t docker.discovery

**Comment by melissacrawford396** [comment]
[root@ip-000000 ~]# ls -lah /sys/fs/cgroup/memory/system.slice/
total 0
drwxr-xr-x. 6 root root 0 Oct 28 21:18 .
drwxr-xr-x. 3 root root 0 Oct 28 16:32 ..
-rw-r--r--. 1 root root 0 Oct 28 16:41 cgroup.clone_children
--w--w--w-. 1 root root 0 Oct 28 16:41 cgroup.event_control
-rw-r--r--. 1 root root 0 Oct 30 10:56 cgroup.procs
drwxr-xr-x. 2 root root 0 Oct 28 19:17 docker-6db458d3a1190f9c9d2e7c305c718e12a63f98408e847673089c988f75406061.scope
drwxr-xr-x. 2 root root 0 Oct 28 21:54 docker-7e708b9ef4a2a622a8c507be3fd159ab8ad3d6f29c892a2765dbef01516937f1.scope
drwxr-xr-x. 2 root root 0 Oct 28 19:26 docker-ca4f29bab2ea575ba4288acf3a8d8f587d0efd51ef70e7856e176f274a699ed3.scope
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.failcnt
--w-------. 1 root root 0 Oct 28 16:41 memory.force_empty
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.failcnt
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.limit_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.max_usage_in_bytes
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.slabinfo
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.tcp.failcnt
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.tcp.limit_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.tcp.max_usage_in_bytes
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.tcp.usage_in_bytes
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.kmem.usage_in_bytes
-rw-r--r--. 1 root root 0 Oct 30 11:17 memory.limit_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.max_usage_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.memsw.failcnt
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.memsw.limit_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.memsw.max_usage_in_bytes
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.memsw.usage_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.move_charge_at_immigrate
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.numa_stat
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.oom_control
----------. 1 root root 0 Oct 28 16:41 memory.pressure_level
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.soft_limit_in_bytes
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.stat
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.swappiness
-r--r--r--. 1 root root 0 Oct 28 16:41 memory.usage_in_bytes
-rw-r--r--. 1 root root 0 Oct 28 16:41 memory.use_hierarchy
drwxr-xr-x. 2 root root 0 Oct 28 17:11 mesos-slave.service
-rw-r--r--. 1 root root 0 Oct 28 16:41 notify_on_release
-rw-r--r--. 1 root root 0 Oct 28 16:41 tasks


[root@ip-000000 ~]# ls -lah /sys/fs/cgroup/cpu,cpuacct/system.slice/
total 0
drwxr-xr-x. 55 root root 0 Oct 30 10:56 .
drwxr-xr-x.  4 root root 0 Oct 28 16:32 ..
drwxr-xr-x.  2 root root 0 Oct 28 16:41 brandbot.service
-rw-r--r--.  1 root root 0 Oct 28 16:41 cgroup.clone_children
--w--w--w-.  1 root root 0 Oct 28 16:41 cgroup.event_control
-rw-r--r--.  1 root root 0 Oct 28 16:41 cgroup.procs
drwxr-xr-x.  2 root root 0 Oct 28 16:41 chef-client.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 choose_repo.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 cloud-config.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 cloud-final.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 cloud-init-local.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 cloud-init.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 cphalod.service
-r--r--r--.  1 root root 0 Oct 28 16:41 cpuacct.stat
-rw-r--r--.  1 root root 0 Oct 28 16:41 cpuacct.usage
-r--r--r--.  1 root root 0 Oct 28 16:41 cpuacct.usage_percpu
-rw-r--r--.  1 root root 0 Oct 28 16:41 cpu.cfs_period_us
-rw-r--r--.  1 root root 0 Oct 28 16:41 cpu.cfs_quota_us
drwxr-xr-x.  2 root root 0 Oct 28 16:41 cpupower.service
-rw-r--r--.  1 root root 0 Oct 28 16:41 cpu.rt_period_us
-rw-r--r--.  1 root root 0 Oct 28 16:41 cpu.rt_runtime_us
-rw-r--r--.  1 root root 0 Oct 30 11:17 cpu.shares
-r--r--r--.  1 root root 0 Oct 28 16:41 cpu.stat
drwxr-xr-x.  2 root root 0 Oct 28 16:41 crond.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 data.mount
drwxr-xr-x.  2 root root 0 Oct 28 16:41 dbus.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 dbus.socket
drwxr-xr-x.  2 root root 0 Oct 28 19:17 docker-6db458d3a1190f9c9d2e7c305c718e12a63f98408e847673089c988f75406061.scope
drwxr-xr-x.  2 root root 0 Oct 28 21:54 docker-7e708b9ef4a2a622a8c507be3fd159ab8ad3d6f29c892a2765dbef01516937f1.scope
drwxr-xr-x.  2 root root 0 Oct 28 19:26 docker-ca4f29bab2ea575ba4288acf3a8d8f587d0efd51ef70e7856e176f274a699ed3.scope
drwxr-xr-x.  2 root root 0 Oct 28 16:41 docker.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 docker-storage-setup.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 encase.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 irqbalance.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 kdump.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 mesos-master.service
drwxr-xr-x.  2 root root 0 Oct 28 17:11 mesos-slave.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 microcode.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 -.mount
drwxr-xr-x.  2 root root 0 Oct 28 16:41 netconsole.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 NetworkManager.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 NetworkManager-wait-online.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 network.service
-rw-r--r--.  1 root root 0 Oct 28 16:41 notify_on_release
drwxr-xr-x.  2 root root 0 Oct 28 16:41 ntpdate.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 ntpd.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 opt-docker.mount
drwxr-xr-x.  2 root root 0 Oct 28 16:41 opt-shipyard.mount
drwxr-xr-x.  2 root root 0 Oct 28 16:41 polkit.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 postfix.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 rc-local.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 rh-cloud-firstboot.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 rhel-dmesg.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 rhnsd.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 rsyslog.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 splunk.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 sshd-keygen.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 sshd.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 sshd.socket
drwxr-xr-x.  2 root root 0 Oct 28 16:41 systemd-ask-password-wall.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 systemd-logind.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 systemd-user-sessions.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 system-getty.slice
drwxr-xr-x.  2 root root 0 Oct 28 16:41 system-serial\x2dgetty.slice
drwxr-xr-x.  2 root root 0 Oct 28 16:41 system-systemd\x2dfsck.slice
-rw-r--r--.  1 root root 0 Oct 28 16:41 tasks
drwxr-xr-x.  2 root root 0 Oct 28 16:41 tuned.service
drwxr-xr-x.  2 root root 0 Oct 29 13:28 zabbix-agent.service
drwxr-xr-x.  2 root root 0 Oct 28 16:41 zookeeper.service


[root@ip-00000 ~]# zabbix_agentd -t docker.discovery
docker.discovery                              [s|{"data":[{"{#FCONTAINERID}":"7e708b9ef4a2a622a8c507be3fd159ab8ad3d6f29c892a2765dbef01516937f1","{#HCONTAINERID}":"mesos-20151028-162349-3159214090-8085-5738-S2.c5c80b77-0d40-4447-86fb-42f769318b6d"},{"{#FCONTAINERID}":"ca4f29bab2ea575ba4288acf3a8d8f587d0efd51ef70e7856e176f274a699ed3","{#HCONTAINERID}":"mesos-20151028-162349-3159214090-8085-5738-S2.2b92abb3-99fc-447a-8ee8-103e44b8e9d6"},{"{#FCONTAINERID}":"6db458d3a1190f9c9d2e7c305c718e12a63f98408e847673089c988f75406061","{#HCONTAINERID}":"mesos-20151028-162349-3159214090-8085-5738-S2.5a8e14c8-b3db-47ab-8560-4557d77f8a0a"}]}]
zabbix_agentd [1555]: Warning: /var/run/netns/: [2] No such file or directory
zabbix_agentd [1555]: Warning: uninitialization failed

**Comment by jangaraj** [comment]
It's OK now - only 3 containers are discovered. How to replicate your previous problem?

**Comment by melissacrawford396** [comment]
The initial containers are the only ones that showed the system.slice. Now the docker containers show the correct container id...just for a container that doesn't exist anymore. If a container terminates, it takes my front-end 24 hours to delete it from being listed.

I'm not sure how to replicate the initial system.slice issue. The first containers that were run, didn't function correctly on my end, so I'm thinking they gave bad data to your module.

I'll close this issue now, and thank you for your prompt responses!

**Comment by jangaraj** [comment]
Discovery rule is part of template and frequency of this discovery rule affect how quickly are changes detected in Zabbix. Default frequency is 600 seconds.

In your case was discovered also container with id "system.slice", which is wrong. If you see this problem again, then please reopen issue and provide outputs:

    ls -lah /sys/fs/cgroup/memory/system.slice/
    ls -lah /sys/fs/cgroup/cpu,cpuacct/system.slice/
    zabbix_agentd -t docker.discovery

Could you provide now details about your OS and:

    docker -v
    docker info

Thank you

**Comment by melissacrawford396** [comment]
We are using the docker module on 5 AWS EC2 instances.

docker -v
```Docker version 1.6.2, build ba1f6c3/1.6.2```
docker info
```Containers: 3
Images: 197
Storage Driver: devicemapper
 Pool Name: docker-253:2-262145-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 7.526 GB
 Data Space Total: 107.4 GB
 Data Space Available: 18.7 GB
 Metadata Space Used: 9.253 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.138 GB
 Udev Sync Supported: true
 Data loop file: /opt/docker/devicemapper/devicemapper/data
 Metadata loop file: /opt/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Kernel Version: 3.10.0-229.14.1.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.1 (Maipo)
CPUs: 2
Total Memory: 6.897 GiB
Name: ip-10-200-77-124.kdc.capitalone.com
ID: XKS6:CLZS:SPXY:ZSP7:LLLY:WTLL:O57O:F6CU:APQ6:IMVY:QRM5:DBZL```
